### PR TITLE
Bookmark map with player titles displayed

### PR DESCRIPTION
**Pull request by IhateTrains**
image

### CODE REVIEW

**Comment by IohannesIohannium** [comment]
Marvelous. Small fixes that could be implemented:
- the country colour is too saturated. CK3 maps have it a bit more "blended" with the map itself
- The characters should be floating in their countries. Maybe finding the "central" pixel would work

**Comment by IohannesIohannium** [comment]
Could you also test it with an observer save?

**Comment by IhateTrains** [comment]
> * the country colour is too saturated. CK3 maps have it a bit more "blended" with the map itself

image


> * The characters should be floating in their countries. Maybe finding the "central" pixel would work

image


> * Coloured wastelands would remove some holes. In Darius' case, the Zagros mountains should still be red

image

**Comment by IhateTrains** [comment]
> Could you also test it with an observer save?

Observer save seems to still have a `played_country` entry, and the entry has the ID of the country you selected before going observer mode. This is a result of an observer save conversion:
image

**Comment by IhateTrains** [comment]
OK, should be ready for review.

**Comment by IhateTrains** [comment]
Final result:

https://user-images.githubusercontent.com/29546927/135722918-3c1c73b0-d9f0-4797-800f-91fda8436f34.mp4

File: ImperatorToCK3.UnitTests/CK3/Map/MapDataTests.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+			Assert.Empty(data.NeighborsDict);
+		}
+		[Fact]
+		public void NeighborProvincesCanBeDetermined() {
+			const string testCK3Path2 = "TestFiles/CK3_all_prov_defs";
+			const ulong byzantionId = 496;
+			var provincesMap = new ImageMagick.MagickImage(testCK3Path2 + "/game/map_data/provinces.png");
+			var definitions = new ProvinceDefinitions(testCK3Path2);
+			Assert.True(definitions.ProvinceToColorDict.ContainsKey(496));
```
Comment (Idhrendur) [inline ImperatorToCK3.UnitTests/CK3/Map/MapDataTests.cs:26]: You defined byzantionId, may as well use it here.

File: ImperatorToCK3.UnitTests/CK3/Map/MapDataTests.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+			var data = new MapData(provincesMap, definitions, testCK3Path);
+			Assert.True(data.NeighborsDict.ContainsKey(496));
```
Comment (Idhrendur) [inline ImperatorToCK3.UnitTests/CK3/Map/MapDataTests.cs:29]: and here

File: ImperatorToCK3/Outputter/BookmarkOutputter.cs [hunk 3]
```diff
@@ -1,30 +1,51 @@
...
 	public static class BookmarkOutputter {
-		public static void OutputBookmark(string outputModName, Dictionary<string, Character> characters, Dictionary<string, Title> titles, Date ck3BookmarkDate) {
-			var path = "output/" + outputModName + "/common/bookmarks/00_bookmarks.txt";
+		public static void OutputBookmark(Dictionary<string, Character> characters, Dictionary<string, Title> titles, Configuration config) {
+			OpenCL.IsEnabled = true; // enable OpenCL in ImageMagick
+			var path = "output/" + config.OutputModName + "/common/bookmarks/00_bookmarks.txt";
 			using var stream = File.OpenWrite(path);
 			using var output = new StreamWriter(stream, Encoding.UTF8);
 
+			// get province positions
```
Comment (Idhrendur) [inline ImperatorToCK3/Outputter/BookmarkOutputter.cs:20]: On a design level, it probably makes more sense to not input things in an outputter. Have CK3 world import them and pass them in here. You'll probably find uses for them elsewhere anyways.

**Review by Idhrendur — APPROVED** [review]

**Pull request merged**